### PR TITLE
Upon successful meeting note creation, socket message is emitted to the associated room

### LINKED ISSUE #175: Sockets in a meeting room should receive a message when a new note is created
## Expected behaviour

When a new meeting note is created by the `POST /meetings/:id/notes` route handler, after the note is successfully saved to the database, a `meeting_note:created` message in the following format is emitted to the Socket.IO room associated with the meeting, such as `meeting:123`.

```json
{
  "id": 1,
  "agenda_owning_participant_id": 2,
  "authoring_participant_id": 3,
  "sort_order": 4,
  "note_text": "The text of the note.",
  "created_at": "2022-02-22T22:22:22.222Z"
}
```

## Actual behaviour

No messages are emitted on the socket when new meeting notes are created.

## Details and resources

There is nothing listening for this message type yet. There aren't even any sockets in the meeting room. This work can be done in parallel to the tickets that add those features.

## Checklist

- [x] This issue is linked to the appropriate project.
- [x] This issue has appropriate labels.

### PR DESCRIPTION
**Pull request by eileenxue**
## Proposed changes

- When a meeting note is successfully created, it will emit a socket message to the room associated with the meeting. 
- Add test cases in `meetingService` to test for new query data and null values

image
Resolves #175 

## Checklist

- [x] Are the issues being addressed linked to this PR?
- [x] Do all commit messages start with the issue number?
- [x] Are all code changes sufficiently tested?
- [x] Are there screenshots for UI changes?

### CODE REVIEW

File: api/src/middleware/routerTestUtils.ts [hunk 1]
```diff
@@ -12,14 +15,26 @@ export const sessionDestroyMock = jest.fn();
 
 export const createAppAgentForRouter = (router: Router): SuperAgentTest => {
   const app = express();
+
+  const server = createServer(app);
+  const io = new Server(server, {
+    cors: {
```
Comment (davidvandusen) [inline api/src/middleware/routerTestUtils.ts:21]: The test server doesn't need CORS configuration since all the requests are made by tests, not browsers.

File: api/src/middleware/routerTestUtils.ts [hunk 2]
```diff
@@ -12,14 +15,26 @@ export const sessionDestroyMock = jest.fn();
...
+      origin: [findConfig('WEBAPP_ORIGIN', '')],
+      credentials: true,
+    },
+  });
+
+  const sessionMiddleware = expressSession({
```
Comment (davidvandusen) [inline api/src/middleware/routerTestUtils.ts:27]: Why was `sessionMiddleware` extracted to a variable?

File: api/src/middleware/routerTestUtils.ts [hunk 3]
```diff
@@ -12,14 +15,26 @@ export const sessionDestroyMock = jest.fn();
 
 export const createAppAgentForRouter = (router: Router): SuperAgentTest => {
   const app = express();
+
+  const server = createServer(app);
```
Comment (davidvandusen) [inline api/src/middleware/routerTestUtils.ts:19]: A server is created from the `app`, but `agent` is being called with `app`, not `server` below. Does this mean that the app agent won't actually handle attempted socket connections?

Does this server need to be created? Would a mock of `io` be sufficient for testing instead?

File: api/src/services/__tests__/meetingsService.ts [hunk 4]
```diff
@@ -152,6 +152,11 @@ describe('meetingsService', () => {
         [agendaOwningParticipantId, 6, noteText, sortOrder],
         [meetingNoteId]
       );
+      mockQuery(
+        'select `id`, `agenda_owning_participant_id`, `authoring_participant_id`, `sort_order`, `note_text`, `created_at` from `meeting_notes` where `id` = ?',
+        [meetingNoteId],
+        [{ id: 5 }]
```
Comment (davidvandusen) [inline api/src/services/__tests__/meetingsService.ts:158]: The mocked SQL and this response don't match. The query selects six fields, but only `id` is in the object in the response.

File: api/src/services/__tests__/meetingsService.ts [hunk 5]
```diff
@@ -162,7 +167,106 @@ describe('meetingsService', () => {
         )
       ).toEqual({ id: meetingNoteId });
     });
-
+    it('should resolve to null if no meeting note id exists', async () => {
```
Comment (davidvandusen) [inline api/src/services/__tests__/meetingsService.ts:170]: I don't understand what situation this description is referring to. It looks like it's the situation where the insert statement doesn't insert a new row. Is there a better way to communicate that to the reader?

File: api/src/services/__tests__/meetingsService.ts [hunk 6]
```diff
@@ -162,7 +167,106 @@ describe('meetingsService', () => {
...
+        await createMeetingNote(
+          meetingId,
+          principalId,
+          agendaOwningParticipantId,
+          noteText,
+          sortOrder
+        )
+      ).toEqual(null);
+    });
+    it('should return a meetingNote object from a given meeting note id', async () => {
```
Comment (davidvandusen) [inline api/src/services/__tests__/meetingsService.ts:196]: This test description doesn't seem to describe the test case. The test case calls `createMeetingNote` and asserts that a newly created meeting note is returned. There's already a test case for that on line [138](https://github.com/OpenTree-Education/rhizone-lms/pull/336/files#diff-685314a9340c322ec0afa4fcb0607baa7c735cf90e3892e67dbf72c54686eb80R138) above. Is this test redundant?

File: api/src/middleware/routerTestUtils.ts (hunk 1 above)
Comment (eileenxue) [reply api/src/middleware/routerTestUtils.ts:21]: Okay, I will remove the CORS configs in this file.

File: api/src/middleware/routerTestUtils.ts (hunk 2 above)
Comment (eileenxue) [reply api/src/middleware/routerTestUtils.ts:27]: Looking at the code again and comparing it with the server code, we didn't actually need to extract it to a new variable since `sessionMiddleware` is not being used anywhere else in the test.

File: api/src/middleware/routerTestUtils.ts (hunk 3 above)
Comment (eileenxue) [reply api/src/middleware/routerTestUtils.ts:19]: I fixed it so that the agent is calling the server. Yes, the server will need to be created in this situation so that it can be used to test socket connections.

Comment (davidvandusen) [reply api/src/middleware/routerTestUtils.ts:19]: Sounds good. 👍

File: api/src/services/__tests__/meetingsService.ts (hunk 4 above)
Comment (eileenxue) [reply api/src/services/__tests__/meetingsService.ts]: Okay, you're right. I added additional variables to represent the `meetingNote` object which is also the response of that query. Then I updated the expected response for the test to also equal `meetingNote`.

File: api/src/services/__tests__/meetingsService.ts (hunk 6 above)
Comment (eileenxue) [reply api/src/services/__tests__/meetingsService.ts]: Yes this test case is redundant after fixing the previous test case. This has now been removed.

File: api/src/services/__tests__/meetingsService.ts (hunk 5 above)
Comment (eileenxue) [reply api/src/services/__tests__/meetingsService.ts]: Yes, there's a better way to communicate that to the reader. When the meeting note responds with an empty array, it should resolve to null.

File: api/src/middleware/meetingsRouter.ts [hunk 7]
```diff
@@ -105,6 +105,9 @@ meetingsRouter.post('/:id/notes', async (req, res, next) => {
     );
     return;
   }
+
+  req.io.to(`meeting:${meetingId}`).emit('meeting_note:created', meetingNote);
```
Comment (pthan1) [inline api/src/middleware/meetingsRouter.ts:109]: Why is `req` used instead of `res` here?

**Review by pthan1 — APPROVED** [review]
From the screenshot, it looks like the code is working.  Are there any console errors?

**Review by davidvandusen — APPROVED** [review]

File: api/src/middleware/meetingsRouter.ts (hunk 7 above)
Comment (eileenxue) [reply api/src/middleware/meetingsRouter.ts:109]: We're using req here because it's trying to send/emit a message out to the specified room, it's not responding to a request.

Comment (pthan1) [reply api/src/middleware/meetingsRouter.ts:109]: That makes sense.  `io` was also placed in the `req` object in the `meetingsService`.

**Comment by eileenxue** [comment]
@pthan1 I don't see any console errors when I'm testing it.
